swe_roe/ex2b.c

```diff
@@ -630,51 +664,97 @@ PetscErrorCode RDyComputeAdditionalEdgeAttributes(DM dm, RDyMesh *mesh) {
+
+    In DMPlex, the cross product of the normal vector to the edge and vector joining the
+    vertices of the edge (i.e. v1Tov2)  always points in the positive z-direction.
+    However, the vector joining the left and the right cell may not be in the same direction
+    as the normal vector to the edge (Case-2). Thus, the edge information in the Case-2 is
+    updated by spawing the vertex ids and flipping the edge normal.
+    */
+
+    PetscInt v_offset = iedge * 2;
+    PetscInt vid_1    = edges->vertex_ids[v_offset + 0];
+    PetscInt vid_2    = edges->vertex_ids[v_offset + 1];
+
+    RDyVector edge_parallel;  // a vector parallel along the edge in 2D
+    for (PetscInt idim = 0; idim < 2; idim++) {
+      edge_parallel.V[idim] = vertices->points[vid_2].X[idim] - vertices->points[vid_1].X[idim];
+    }
+    edge_parallel.V[2] = 0.0;
 
-    } else if (PetscAbs(edges->normals[iedge].V[1]) < 1.e-10) {
-      // It is a horizontal edge, so
-      // sn = 0.0 and cn = +/- 1.0
-      if (is_internal_edge) {
-        PetscReal xr     = cells->centroids[r].X[0];
-        PetscReal xl     = cells->centroids[l].X[0];
-        PetscReal dx_l2r = xr - xl;
-        if (dx_l2r < 0.0) {
-          edges->cn[iedge] = -1.0;
-        } else {
-          edges->cn[iedge] = 1.0;
-        }
-      } else {
-        edges->cn[iedge] = 1.0;
+    // In case of an internal edge, a vector from the left cell to the right cell.
+    // In case of a boundary edge, a vector from the left cell to edge centroid.
+    // Note: This is a vector in 2D.
+    RDyVector vec_L2RorEC;
+
+    if (is_internal_edge) {
+      for (PetscInt idim = 0; idim < 2; idim++) {
+        vec_L2RorEC.V[idim] = cells->centroids[r].X[idim] - cells->centroids[l].X[idim];
       }
 
     } else {
-      printf("The code only support quad cells with edges that align with x and y axis\n");
-      exit(0);
+      for (PetscInt idim = 0; idim < 2; idim++) {
+        vec_L2RorEC.V[idim] = (vertices->points[vid_2].X[idim] + vertices->points[vid_1].X[idim]) / 2.0 - cells->centroids[l].X[idim];
+      }
     }
+    vec_L2RorEC.V[2] = 0.0;
+
+    // Compute a vector perpendicular to the edge_parallel vector via a clockwise
+    // 90 degree rotation
+    RDyVector edge_perp;
+    edge_perp.V[0] = edge_parallel.V[1];
+    edge_perp.V[1] = -edge_parallel.V[0];
+
+    // Compute the dot product to check if vector joining L-to-R is pointing
+    // in the direction of the vector perpendicular to the edge.
+    PetscReal dot_prod = vec_L2RorEC.V[0] * edge_perp.V[0] + vec_L2RorEC.V[1] * edge_perp.V[1];
+
+    if (dot_prod < 0.0) {
+      // The angle between edge_perp and vec_L2RorEC is greater than 90 deg.
+      // Thus, flip vertex ids and the normal vector
+      edges->vertex_ids[v_offset + 0] = vid_2;
+      edges->vertex_ids[v_offset + 1] = vid_1;
+      for (PetscInt idim = 0; idim < 3; idim++) {
+        edges->normals[iedge].V[idim] = -edges->normals[iedge].V[idim];
```

Comment:
You can also use `edges->normals[iedge].V[idim] *= -1.0;` here to flip the sign.